signature-crate/src/error.rs

```diff
@@ -1,25 +1,86 @@
 //! Signature error types
 
-use core::fmt;
+use core::fmt::{self, Display};
 
-/// Errors which can occur during signature operations
-#[derive(Copy, Clone, Eq, PartialEq, Debug)]
-pub enum Error {
-    /// Signature failed to verify
-    SignatureInvalid,
+#[cfg(feature = "std")]
+use std::{boxed::Box, error::Error as StdError};
+
+/// Signature errors
+#[derive(Debug, Default)]
+pub struct Error {
+    /// Cause of the error (if applicable)
+    #[cfg(feature = "std")]
+    cause: Option<Box<dyn StdError>>,
```

Comment:
An alternative to using `std::error::Error` is the [failure](https://github.com/rust-lang-nursery/failure) crate. However, while I'm a fan of this crate, I'm not sure it buys us anything here, or if it makes sense for all downstream crates to use failure.

Reply:
IIRC `failure` is compatible with the `Error` trait, so I think it's better to leave it as-is. Backtraces will become less detailed, but I think it's fine.